Client App Auth doc (#122)
plentymarkets/plenty-developers-docs: docs/en-gb/modules/rest-api-guides/pages/client-app-auth.adoc

|*Access rights*
|What REST API areas can the client app access
|===

An app client secret is automatically generated when creating the App client.

image::client-app-auth/app-client-secret.png[width=489]

[WARNING]
====
After saving a popup appears with the secret of the app client.
Make sure to copy the *secret*, because you won’t have access to it afterwards.
In case you lose your app client *secret* you have to reset it.
====

After creating the App Client, you will be redirected to the `App Client Details Page`

=== App Client Details Page

Once an App Client is created, you can view the `Client ID` and `Token URL` used for generating the JWT token.

The `Details Page` has the following options:

|===
|Option |Explanation

|*Name*
|The name of the App Client used for future reference and identification

|*Client ID*
|Client ID used in the Token URL payload

|*Token URL*
|URL used to generate the JWT token used for REST API authentication

|*Enable app client*

`on/off`
|Enable or disable the client app overall access

|*Rights*
|What REST API endpoints and actions can the client app access

|*Access rights*
|What REST API areas can the client app access

> 

Reviewer: ```suggestion
```

|===

> === Regenerate secret

Reviewer: ```suggestion
=== Regenerate App Client secrets
```

> If you forgot the app client secret or just want to regenerate it you can do this by pressing the `Regenerate secret`

Reviewer: ```suggestion
If you forgot the App Client secret or just want to regenerate it you can do this by clicking on *Regenerate secret*:
```

> button:

Reviewer: ```suggestion

```

image::client-app-auth/regenerate-secret.png[width=280]

[WARNING]
====

> After pressing the button a popup appears with the secret of the app client.

Reviewer: ```suggestion
After clicking the button a popup appears with the *App Client secret*.
```

> Make sure to copy the *secret*, because you won’t have access to it afterwards.

Reviewer: ```suggestion
Copy the *App Client secret*, because you won’t have access to it afterwards.
```

====

== How to authenticate with your Client App

> Authentication in the REST API with an Client App requirest 2 steps:

Reviewer: ```suggestion
Authentication in the REST API with an Client App requires two steps:
```

1. Generate the JWT token from the `Token URL` in the Client App settings
2. Use the JWT token to authenticate the REST API calls

=== Generate JWT Token

Make a `POST` request to the `Token URL` as in the following example:

[source]
----
POST [TOKEN_URL]
Content-Type: application/x-www-form-urlencoded
Payload:
{
   client_id: [CLIENT_ID],
   client_secret: [CLIENT_SECRET],
   grant_type: client_credentials,
}
----

|===
|Variable |Explanation

|TOKEN_URL
|The `TOKEN URL` from the Client App Details Page

|CLIENT_ID
|The `CLIENT ID` from the Client App Details Page

|CLIENT_SECRET
|The latest Client App `secret` that was generated

|===

Success response:
[source]
----
{
    "access_token": [JWT_TOKEN],
    "expires_in": [JWT_TOKEN_EXPIRATION],
    "token_type": "Bearer"
}
----
|===

> |Variable |Explanation

Reviewer: Missing periods at the end of the sentences

|JWT_TOKEN
|The JWT token that can be used to authenticate the REST API calls

|JWT_TOKEN_EXPIRATION
|The amount of seconds that it will take until the JWT will expire from the time the JWT it is generated

The default value is `3600` seconds, but it may vary.

> 

Reviewer: ```suggestion
```

|===

[WARNING]
====

> Once an JWT token is expired, it can not be used

Reviewer: ```suggestion
Once a JWT token is expired, it can not be used.
```

====

> === Authenticate in the REST API

Reviewer: ```suggestion
=== Authenticating in the REST API
```

For authenticating the REST API with the Client App JWT token you have to sign every request sent to the REST API
endpoints using the `Bearer Token` Authorization Type by setting the `Authorization` header:

----
Authorization: Bearer [JWT_TOKEN]
----

== FAQ

=== Is there a token refresh for the JWT token?

> No, at this time there is no way to refresh the JWT token without requesting a new one from the `TOKEN URL`

Reviewer: ```suggestion
No, at this time there is no way to refresh the JWT token without requesting a new one from the `TOKEN URL`.
```

=== Can you invalidate a JWT token?
No, at this moment we do not provide a way to invalidate a JWT token. A token is valid for `3600` seconds (`1` hour),
after which it will no longer be usable.

In case of emergency, you can disable the Client App until the JWT token expires.

=== How can I find what Client App is authenticated in the REST API?
You can use the `rest/users/me` endpoint to get information about the user for wich the JWT token.

_**Note:** in the json data that you will receive from the `rest/users/me` endpoint, the `CLIENT ID` is found in the

> `username` parameter._

Reviewer: ```suggestion

```
